apps/src/gamelab/GameLab.js

```diff
@@ -826,38 +826,132 @@ GameLab.prototype.onP5ExecutionStarting = function () {
+    if (this.areAnimationsReady_()) {
+      resolve();
     } else {
-      if (this.eventHandlers.preload) {
-        this.log("WARNING: preload() was ignored because setup() was not provided");
-        this.eventHandlers.preload = null;
+      // Watch store changes until all the animations are ready.
+      const unsubscribe = store.subscribe(() => {
+        if (this.areAnimationsReady_()) {
+          unsubscribe();
+          resolve();
+        }
+      });
+    }
+  }).then(() => {
+    // Animations are ready - send them to p5 to be loaded into the engine.
+    this.gameLabP5.preloadAnimations(store.getState().animationList);
+  });
+};
+
+/**
+ * Check whether all animations in the project animation list have been loaded
+ * into memory and are ready to use.
+ * @returns {boolean}
+ * @private
+ */
+GameLab.prototype.areAnimationsReady_ = function () {
+  const animationList = this.studioApp_.reduxStore.getState().animationList;
+  return animationList.orderedKeys.every(key => animationList.propsByKey[key].loadedFromSource);
+};
+
+/**
+ * Run the preload event handler, and optionally global code, and report when
+ * it is done by resolving a returned Promise.
+ * @returns {Promise} Which will resolve immediately if there is no code to run,
+ *          otherwise will resolve when the preload handler has completed.
+ * @private
+ */
+GameLab.prototype.runPreloadEventHandler_ = function () {
+  return new Promise(resolve => {
+    this.initInterpreter();
+    // Execute the interpreter for the first time:
+    if (this.JSInterpreter && this.JSInterpreter.initialized()) {
+      // Start executing the interpreter's global code as long as a setup() method
+      // was provided. If not, we will skip running any interpreted code in the
+      // preload phase and wait until the setup phase.
+      this.reportPreloadEventHandlerComplete_ = () => {
+        this.reportPreloadEventHandlerComplete_ = null;
+        resolve();
+      };
+      if (this.globalCodeRunsDuringPreload) {
+        this.JSInterpreter.executeInterpreter(true);
+        this.interpreterStarted = true;
+
+        // In addition, execute the global function called preload()
+        if (this.eventHandlers.preload) {
+          this.eventHandlers.preload.apply(null);
+        }
+      } else {
+        if (this.eventHandlers.preload) {
+          this.log("WARNING: preload() was ignored because setup() was not provided");
+          this.eventHandlers.preload = null;
+        }
       }
+      this.completePreloadIfPreloadComplete();
+    } else {
+      // If we didn't run anything resolve now.
+      resolve();
     }
-    this.completePreloadIfPreloadComplete();
+  });
+};
+
+/**
+ * Called on tick to check whether preload code is done running, and trigger
+ * the appropriate report of completion if it is.
+ */
+GameLab.prototype.completePreloadIfPreloadComplete = function () {
+  if (typeof this.reportPreloadEventHandlerComplete_ !== 'function') {
```

Comment:
might be worth explicitly mentioning that this function is created by `runPreloadEventHandler_`